### PR TITLE
Support for Views Tripal Entity Relationship

### PR DESCRIPTION
**Pull request by spficklin**
Issue #467 

## Type(s) of Change(s)

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API-specific change (fix or addition to an API function)
- [ ] Updates documentation (inline or markdown files)

## Description

Currently, it's not possible to create relationships between a Tripal entity and other Drupal related tables such as Taxonomy.  So, for example, if you add the "Taxonomy Term Reference" field to any of your entities you will be able to tag them,  but you can never filter your entities by those tags.  This is because our code for entities doesn't support it in the TripalFieldQuery nor in Views.   This PR fixes both.  It also fixes some major problems with integrating results should we ever have other storage backends... which in fact is the Taxonomy example is one. 

I'm marking this as a release blocker because it fixes some major issues with queries.

## Testing?

To test do the following:

NOTE:  please run this test on a non critical machine.  This is because it has a very important update function that  adds a new setting to all Tripal fields to help distinguish them from non tripal fields.  I have tested the update on my machine and it worked fine but I would hate for the test to not work on a production machine.

1)  After pulling this branch you will need to run ```drush updatedb```.
2)  Go the Admin > Structure > Taxonomy and create a new vocabulary named "Publication Tags".
3)  Add some new terms to it.  You can pick anything that makes sense to you to tag publications (e.g. "Software", "Genome", etc.).
4)  Go to your publication content type:  Admin > Structure > Tripal Content Types.  Click the 'manage fields' for the Publication type.  Add a new field of type 'Term Reference'.  Name it 'field_pub_tags'. If you give it that exact name then the view I attach to this PR should work.  The following two screen shots show the settings I think are best for testing but it really shouldn't matter what you choose. tag_field_settings tag_field_settings2
Note: if you enable the term_reference module you get the nice tree view of terms.
5) Click on the 'Manage Display' field to move your new tags field out of the Disabled section and somewhere that you can see it on the page and save.
6) Now search for a publication entity on your site and add a new tag to by editing the publication and setting the tag.
7) Next create a new view that uses the Publication content type.  For a shortcut you can use the exported view here: [pub_tags-view.txt](https://github.com/tripal/tripal/files/2125795/pub_tags-view.txt).  Here's a screenshot of the settings for the view:  pub_tags-view
8)  Now use the view to search for the publication you tagged in step 6.  This view lets you test filters on both Tripal fields (e.g. publcation year) and Taxonomy tags (Drupal field).  You should be able successfully use any combination of both to filter your pubs.

## Additional Notes (if any):

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by bradfordcondon** [comment]
For those who follow instructions too literally: Name your field `pub_tag`, **not** `field_pub_tags` (@spficklin you may want to edit instructions) to use the view that stephen has very kindly provided.  (the resulting machine name will be field_pub_tags... i ended up with a field named `field_field_pub_tags` and had to destroy it/start over).

File: tripal/includes/TripalFieldQuery.inc [hunk 1]
```diff
@@ -32,49 +90,109 @@ class TripalFieldQuery extends EntityFieldQuery {
...
+    // This is the first result set.
+    if (array_key_exists('first_results', $current_results)) {
+      return $new_results;
+    }
+    
+    // Iterate through all of the new results and only include those that
+    // exist in both the current and new.
+    $intersection = [];
+    foreach ($new_results['TripalEntity'] as $entity_id => $stub) {
+      if (array_key_exists($entity_id, $current_results['TripalEntity'])) {
```
Comment (bradfordcondon) [inline tripal/includes/TripalFieldQuery.inc:156]: if i do a search that returns no results (IE pick a pub year that is wrong) I get the following errors at this line:


```
Notice: Undefined index: TripalEntity in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Warning: array_key_exists() expects parameter 2 to be array, null given in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Notice: Undefined index: TripalEntity in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Warning: array_key_exists() expects parameter 2 to be array, null given in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
```

**Review by bradfordcondon — CHANGES_REQUESTED** [review]
Functionality confirmed!  Filtering by tags works.  It also works in combination with filtering by tripal field.

Here's the view succeeding:

screen shot 2018-06-28 at 9 06 45 am

But what happens if we return no results?

screen shot 2018-06-28 at 9 06 17 am

We should probably fix the error for null results on line 148.

**Comment by bradfordcondon** [comment]
Note: it gracefully handles the TAG not existing.

screen shot 2018-06-28 at 9 07 28 am

searching for a tag that has no records gives no error (but also no notice... i dont know if thats this PR's job)
screen shot 2018-06-28 at 9 08 12 am

**Comment by bradfordcondon** [comment]
PR made to fix the issues I brought up above.  https://github.com/tripal/tripal/pull/488
Apologies if you'd prefer I commit directly...

**Comment by spficklin** [comment]
@bradfordcondon I've merged your PR into this one.  Are you satisfied now that this is merged in? Perhaps @shawnawsu can provide a second review now.

**Review by bradfordcondon — APPROVED** [review]
thanks!

**Comment by shawnawsu** [comment]
Just finished test PR #453, will now test this one.

**Comment by shawnawsu** [comment]
I test this and it works for me including the updated graceful fail. So I approve this, will add the tag.

**Pull request merged**